ui/server/src/main/scala/pl/touk/nussknacker/ui/notifications/NotificationService.scala

```diff
@@ -0,0 +1,104 @@
+      case Success(Some(scenarioName)) => synchronized {
+        data.append(NotificationEvent(UUID.randomUUID().toString, event, now, user, scenarioName))
+      }
+    }
+    filterOldNotifications(now)
+  }
+
+  private def filterOldNotifications(now: Instant): Unit = synchronized {
+    data.zipWithIndex.filter(_._1.date.isBefore(now.minus(config.duration.toMillis, ChronoUnit.MILLIS))).foreach(i => data.remove(i._2))
+  }
+
+  private[notifications] def dataFor(user: LoggedUser, notificationsAfter: Option[Instant]): List[NotificationEvent] = {
+    filterOldNotifications(Instant.now(clock))
+    data.filter(event => event.user.id == user.id && !notificationsAfter.exists(_.isAfter(event.date))).toList
+  }
+
+
+}
+
+trait CurrentDeployments {
+  def retrieve(implicit timeout: Timeout): Future[DeploymentStatusResponse]
+}
+
+class ManagementActorCurrentDeployments(managementActor: ActorRef) extends CurrentDeployments {
+  override def retrieve(implicit timeout: Timeout): Future[DeploymentStatusResponse] = (managementActor ? DeploymentStatus).mapTo[DeploymentStatusResponse]
+}
+
+class NotificationService(currentDeployments: CurrentDeployments,
+                          store: NotificationsListener) {
+
+  def notifications(user: LoggedUser, notificationsAfter: Option[Instant])(implicit ec: ExecutionContext, timeout: Timeout): Future[List[Notification]] = {
+    Future.sequence(List(
+      prepareDeploymentNotifications(user),
+      Future.successful(userDeployments(user, notificationsAfter))
+    )).map(_.flatten)
+  }
+
+  private def userDeployments(user: LoggedUser, notificationsAfter: Option[Instant]): Seq[Notification] = {
+    store.dataFor(user, notificationsAfter).collect {
+      case NotificationEvent(id, OnDeployActionFailed(_, reason), _, _, name) =>
+        Notification(id, Some(name), s"Deployment of ${name.value} failed with ${reason.getMessage}", NotificationType.error,
+          List(DataToRefresh.versions, DataToRefresh.activity))
+      case NotificationEvent(id, _: OnDeployActionSuccess, _, _, name) =>
+        Notification(id, Some(name), s"Deployment finished", NotificationType.success, List(DataToRefresh.versions, DataToRefresh.activity))
+    }
+  }
+
+  private def prepareDeploymentNotifications(user: LoggedUser)(implicit ec: ExecutionContext, timeout: Timeout): Future[List[Notification]] = {
+    currentDeployments.retrieve.map { case DeploymentStatusResponse(deploymentInfos) =>
+      deploymentInfos
+        //no need to inform current user
+        .filterNot(_._2.userId == user.id)
+        .map { case (k, v) => currentDeploymentToNotification(k, v) }.toList
+    }
+  }
+
+  private def currentDeploymentToNotification(processName: ProcessName, deploymentInfo: DeployInfo): Notification = {
+    val actionString = deploymentInfo.action match {
+      case DeploymentActionType.Deployment => "deployed"
+      case DeploymentActionType.Cancel => "cancelled"
+    }
+    //TODO: should it be displayed only once?
```

Comment:
meaning right now it's shown multiple times ?

Reply:
it keeps popping up, until deployment finishes - I've kept current behaviour